From 47632974d986fe848a9718a44dd08f0dbfc361df Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:14 -0400
Subject: [net] openvswitch: Enable tunnel GSO for OVS bridge

Message-id: <566b48ebffa40fe518fd004e676c4137847e3642.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94007
O-Subject: [RHEL7.1 net PATCH 41/50] openvswitch: Enable tunnel GSO for OVS bridge.
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit f6eec614d2252a99b861e288b6301599d2d58da4
Author: Pravin B Shelar <pshelar@nicira.com>
Date:   Thu Jul 17 15:14:15 2014 -0700

    openvswitch: Enable tunnel GSO for OVS bridge.

    Following patch enables all available tunnel GSO features for OVS
    bridge device so that ovs can use hardware offloads available to
    underling device.

    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>
    Acked-by: Andy Zhou <azhou@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/netdev_features.h b/include/linux/netdev_features.h
index c8c838d..799c605 100644
--- a/include/linux/netdev_features.h
+++ b/include/linux/netdev_features.h
@@ -173,4 +173,12 @@ enum {
 /* changeable features with no special hardware requirements */
 #define NETIF_F_SOFT_FEATURES (NETIF_F_GSO | NETIF_F_GRO)
 
+#define NETIF_F_GSO_ENCAP_ALL (NETIF_F_GSO_GRE |   \
+     NETIF_F_GSO_GRE_CSUM |   \
+     NETIF_F_GSO_IPIP |   \
+     NETIF_F_GSO_SIT |   \
+     NETIF_F_GSO_UDP_TUNNEL |  \
+     NETIF_F_GSO_UDP_TUNNEL_CSUM |  \
+     NETIF_F_GSO_MPLS)
+
 #endif /* _LINUX_NETDEV_FEATURES_H */
diff --git a/net/openvswitch/vport-internal_dev.c b/net/openvswitch/vport-internal_dev.c
index 295471a..4e8285f 100644
--- a/net/openvswitch/vport-internal_dev.c
+++ b/net/openvswitch/vport-internal_dev.c
@@ -140,11 +140,14 @@ static void do_setup(struct net_device *netdev)
  netdev->tx_queue_len = 0;
 
  netdev->features = NETIF_F_LLTX | NETIF_F_SG | NETIF_F_FRAGLIST |
-      NETIF_F_HIGHDMA | NETIF_F_HW_CSUM | NETIF_F_GSO_SOFTWARE;
+      NETIF_F_HIGHDMA | NETIF_F_HW_CSUM |
+      NETIF_F_GSO_SOFTWARE | NETIF_F_GSO_ENCAP_ALL;
 
  netdev->vlan_features = netdev->features;
+ netdev->hw_enc_features = netdev->features;
  netdev->features |= NETIF_F_HW_VLAN_CTAG_TX;
  netdev->hw_features = netdev->features & ~NETIF_F_LLTX;
+
  eth_hw_addr_random(netdev);
 }
 
-- 
1.7.1